File: JAVA/14_Backtracking/BOJ_16637_괄호추가하기.java
import java.io.IOException;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;

public class Main {

    static int max = Integer.MIN_VALUE;
    static ArrayList<Integer> number;
    static ArrayList<Character> operator;

    public static int cal(int n1, char op, int n2) {
        if (op == '+') return n1 + n2;
        else if (op == '-') return n1 - n2;
        else return n1 * n2;
    }

    public static void dfs(int dpth, int sum) {
        if (dpth == operator.size()) {
            max = Math.max(max, sum);
            return;
        }
        int tmp1 = cal(sum, operator.get(dpth), number.get(dpth + 1));
        dfs(dpth + 1, tmp1);
        if (dpth + 1 < operator.size()) {
            int tmp2 = cal(number.get(dpth + 1), operator.get(dpth + 1), number.get(dpth + 2));
            tmp2 = cal(sum, operator.get(dpth), tmp2);
            dfs(dpth + 2, tmp2);
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        String str = br.readLine();
        number = new ArrayList<>();
        operator = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            char c = str.charAt(i);
            if (c == '+' || c == '-' || c == '*') {
                operator.add(c);
            } else {
                number.add(c - '0');
            }
        }
        dfs(0, number.get(0));
        System.out.println(max);
    }
}

File: C/14_Backtracking/BOJ_16637_괄호추가하기.c
#include <stdio.h>
#define math_max(a, b) a > b ? a : b
int n, max = -2147483648, *number, op_size = 0, num_size = 0;
char* operator;
int cal(int n1, char op, int n2) {
	if (op == '+') return n1 + n2;
	else if (op == '-') return n1 - n2;
	else return n1 * n2;
}
void dfs(int dpth, int sum) {
	if (dpth == op_size) {
		max = math_max(max, sum);
		return;
	}
	int tmp1 = cal(sum, operator[dpth], number[dpth + 1]);
	dfs(dpth + 1, tmp1);
	if (dpth + 1 < op_size) {
		int tmp2 = cal(number[dpth + 1], operator[dpth + 1], number[dpth + 2]);
		tmp2 = cal(sum, operator[dpth], tmp2);
		dfs(dpth + 2, tmp2);
	}
}
main() {
	scanf("%d", &n);
	number = (int*)malloc(sizeof(int) * (n / 2 + 1));
	operator = (char*)malloc(sizeof(char) * (n / 2));
	char* str = (char*)malloc(sizeof(char) * n);
	scanf("%s", str);
	for (int i = 0; i < n; i++) {
		char c = str[i];
		if (c == '+' || c == '-' || c == '*') {
			operator[op_size++] = c;
		}
		else {
			number[num_size++] = c - '0';
		}
	}
	dfs(0, number[0]);
	printf("%d", max);
}